[FEATURE] 인증 성공, 실패 핸들러 구현

JSY8869 opened the pull request:
## 제목
Jwt기반으로 실제 토큰을 생성하고 제공하는 로직 구현

## 작업 내용
1. jjwt-jackson 라이브러리 추가

jjwt를 사용하기 위해 필요한 라이브러리를 build 파일에 추가

2. 인증 실패 핸들러 구현

모든 인증 실패 핸들러에서 사용하는 로직을 상위 인터페이스의 메서드로 정의하고 jwt 관련 인증 실패 핸들러를 구현

3. 인증 성공 핸들러 구현

모든 인증 성공 핸들러에서 사용하는 로직(default 응답)을 상위 인터페이스의 메서드로 정의하고 jwt 관련 인증 실패 핸들러를 구현

4. handler 관련 configuration 설정

구현한 handler를 spring bean으로 등록하였으며, security configuration에 추가하였음

5. 테스트 중 발생한 오류 해결

테스트 중 발생한 오류 들을 코드 수정을 통해 해결하였으며, 후에 refreshToken 관련 수정이 필요함(쿠키에서 `space`가 `+`로 바뀌는 부분)

JSY8869 committed:
chore: Add jjwt-jackson

1. jjwt를 사용하기 위해 필요한 라이브러리 추가
jjwt를 사용하기 위해 필요한 구현체 라이브러리인 jjwt-jackson을 추가하였습니다.

related to #14

JSY8869 committed:
feat: Add authentication failure handler

1. 인증 실패 handler 상위 인터페이스 구현
authentication failure handler에 공통적으로 들어갈 메서드를 선언하기 위해 상위 인터페이스를 구현하였으며, spring security와 연결하기 위해 spring security의 AuthenticationFailureHandler을 상속 받았습니다.

2. Jwt 인증 실패 handler 구현
각각의 에러에 따라 응답을 동적으로 생성할 수 있도록 setErrorResponse 메서드에서 에러의 형태를 입력 받아 응답을 생성하도록 구현하였습니다.

related to #14

JSY8869 committed:
feat: Add authentication success handler

1. 인증 성공 handler 상위 인터페이스 구현
모든 인증 성공 handler에 들어갈 공통 로직인 Default response를 생성하는 메서드를 선언하기 위해 상위 인터페이스를 구현하였습니다.

2. 인증 성공 handler jwt 인터페이스 구현
jwt의 인증 성공 시 access 토큰과 refresh 토큰을 응답에 포함 시키는  메서드를 선언하기 위해 상위 인터페이스와 구현체 중간에 jwt 인터페이스를 구현하였습니다.

3. 인증 성공 handler jwt 구현
jwt 인증 성공 시 실제로 토큰을 생성하고, 응답에 포함시키는 로직을 구현하였으며, cookie에 refresh 토큰을 저장하기 위해 인코딩이 필요하여 인코딩 로직을 추가하였습니다.

4. AccessToken, RefreshToken getter 추가
인증 성공 후 토큰의 value를 응답에 추가하기 위해 getter를 추가하였습니다.

related to #14

JSY8869 committed:
feat: Add authentication handler configuration

1. jwt 인증 성공, 실패 handler configurate

related to #14

JSY8869 committed:
fix: Resolve errors encountered during testing

1. JwtAuthenticationToken principal 변경
인증 성공 handler에서 클래스 형변환 오류 해결을 위해 principal에 들어가는 정보를 userDetails 객체 자체로 변경

2. TokenProperties의 SECRET_KEY, REFRESH_KEY 변경
글자 수가 64자를 넘어야 secretKey로 정의할 수 있기 때문에 길이를 64자 이상으로 변경

3. Refresh Token 생성 시 prefix 제거
쿠키 특성 상 space를 넣을 수 없기 때문에 space가 포함되어 있는 prefix 제거

4. Token 인터페이스가 Serializable를 상속하도록 변경
토큰의 값을 직렬화 하여 읽기 위해 Serializable을 상속하도록 변경

5. jwt 토큰 생성 시 반환 값 변경
jwt 토큰을 생성할 때 상위 클래스가 아닌 jwt 토큰 클래스를 반환하도록 변경

related to #14

CokeLee777 commented on src/main/java/io/github/cokelee777/springsecurityjwtauth/security/token/domain/Token.java:
```diff
@@ -1,8 +1,9 @@
 import java.util.Date;
 
-public interface Token {
+public interface Token extends Serializable {
```
이거 지금 보니깐 불필요할 것 같은데 지우고 테스트 한 후 괜찮으면 지우고 커밋해주세요~

CokeLee777 reviewed (commented):
나머지는 좋은 것 같습니다. 커맨트 확인 후 커밋한 다음 다시 커맨트 남겨주세요

JSY8869 committed:
refactor: Remove unnecessary extends

1. Token 인터페이스의 불필요한 상속 제거
Serializable을 상속 받지 않아도 동작함을 확인하였기 때문에 제거

related to #14

JSY8869 committed:
fix: Change the order of generating responses

1. 응답 생성 순서 변경
응답 생성 시 default 응답을 생성하는 과정에서 성공 응답을 끝마치기 때문에 헤더와 쿠키에 토큰을 넣지 못하는 것을 테스트 중 발견하여 응답 생성 순서를 변경

related to #14

JSY8869 commented:
테스트 후 Serializabe은 불필요하다고 판단하여 제거하였습니다.
또한, 쿠키와 헤더에 토큰이 제대로 들어가지 않고 OK 응답을 하는 것을 발견하여 문제 분석 후 성공 응답 생성 시 순서에 문제가 있는 것을 발견하여 순서 변경을 통해 해결하였습니다.

CokeLee777 commented:
> 테스트 후 Serializabe은 불필요하다고 판단하여 제거하였습니다. 또한, 쿠키와 헤더에 토큰이 제대로 들어가지 않고 OK 응답을 하는 것을 발견하여 문제 분석 후 성공 응답 생성 시 순서에 문제가 있는 것을 발견하여 순서 변경을 통해 해결하였습니다.

확인했습니다~ 머지 해주셔도 괜찮을 듯 합니다.
